api/users.rb

```diff
@@ -180,7 +185,12 @@ def set_template_result(user_id, user_stats, thread_ids)
     end
 
     # Get the number of threads read by each user.
-    users = User.only([:_id, :read_states]).where("read_states.course_id" => course_id)
+    users_read_states_selector = {"read_states.course_id" => course_id}
+    if user_ids
+        users_read_states_selector[:_id.in] = user_ids
+    end
+
+    users = User.only([:_id, :read_states]).where(users_read_states_selector)
```

Comment:
I think you want to find a way to move the database query on line 198 out of the loop over `users` below.